### PR TITLE
BlueprintInput deepcopies inputs at initialization

### PR DESCRIPTION
**Pull request by vinferrer**
Closes #314 

As described in #314 ```check_trigger_amounts`` method inside the ``BlueprintInput`` class was modifying the attributes of the objects and at the same time modifying the inputs to the class initialization. Now at the initialization the inputs are deepcopied to avoid these problems 

## Proposed Changes

  - Deepcopy ``BlueprintInput`` inputs at initialization
  - Add assert to test this
  - Eliminate deepcopies from the testing 

## Change Type

- [x] `bugfix` (+0.0.1)
- [ ] `minor` (+0.1.0)
- [ ] `major`  (+1.0.0)
- [ ] `refactoring` (no version update)
- [ ] `test` (no version update)
- [ ] `infrastructure` (no version update)
- [ ] `documentation` (no version update)
- [ ] `other`

## Checklist before review

- [x] I added everything I wanted to add to this PR.
- [x] \[Code or tests only\] I wrote/updated the necessary docstrings.
- [x] \[Code or tests only\] I ran and passed tests locally.
- [ ] \[Documentation only\] I built the docs locally.
- [ ] My contribution is harmonious with the rest of the code: I'm not introducing repetitions.
- [ ] My code respects the adopted style, especially linting conventions.
- [x] The title of this PR is explanatory on its own, enough to be understood as part of a changelog.
- [x] I added or indicated the right labels.

- [ ] I added information regarding the timeline of completion for this PR.

- [x] Please, comment on my PR while it's a draft and give me feedback on the development!

### CODE REVIEW

File: phys2bids/tests/test_physio_obj.py [hunk 1]
```diff
@@ -101,7 +101,7 @@ def test_BlueprintInput():
 
     # Tests rename_channels
     new_names = ['trigger', 'time', 'lindt']
-    blueprint_in.rename_channels(new_names.copy())
+    blueprint_in.rename_channels(new_names)
     assert blueprint_in.ch_name == ['time', 'trigger', 'lindt']
```
Comment (tsalo) [inline phys2bids/tests/test_physio_obj.py:105]:
```suggestion
    assert blueprint_in.ch_name == ['time', 'trigger', 'lindt']
    assert test_index[0] is not test_trigger
```

If it's not storing the original array then this should pass.

**Comment by vinferrer** [comment]
Ah, so it wasn't showing up in the tests before, my mistake

**Comment by eurunuela** [comment]
> Ah, so it wasn't showing up in the tests before, my mistake

It's failing because you're comparing a variable called `text_index` that you haven't declared.

**Comment by tsalo** [comment]
Sorry about that. I couldn't put a comment down where the assert should have gone, but I didn't think the variable hadn't been declared yet where I did comment.

**Comment by vinferrer** [comment]
wait so should it pass once the variable is declared?

**Comment by tsalo** [comment]
If the original variable is stored in the object (i.e., the bug persists), then it will fail. If it isn't then it should pass.

File: phys2bids/tests/test_physio_obj.py [hunk 2]
```diff
@@ -101,11 +101,12 @@ def test_BlueprintInput():
...
     new_names = ['trigger', 'time', 'lindt']
-    blueprint_in.rename_channels(new_names.copy())
+    blueprint_in.rename_channels(new_names)
+    test_index = blueprint_in.return_index(1)
     assert blueprint_in.ch_name == ['time', 'trigger', 'lindt']
+    assert test_index[0] is not test_trigger
 
     # Tests return_index
-    test_index = blueprint_in.return_index(1)
     assert (test_index[0] == test_trigger).all()
```
Comment (vinferrer) [inline phys2bids/tests/test_physio_obj.py:110]: aren't 107 and 110 uncompatible?

Comment (tsalo) [reply phys2bids/tests/test_physio_obj.py:110]: you just need to move 107 down to 110

Comment (vinferrer) [reply phys2bids/tests/test_physio_obj.py:110]: That's not my point, now your line is failing as you said it should. What I am raising as question is why  do you propose an assert (107) that is contradictory to one we already have (110)

Comment (tsalo) [reply phys2bids/tests/test_physio_obj.py]: Ah, I see. `is` isn't the same as `==`. `==` should check that the arrays contain equivalent values (although `np.array_equal` would be cleaner), while `is` should determine if the variables are referring to the same object.

Comment (vinferrer) [reply phys2bids/tests/test_physio_obj.py]: ah, I see your point

**Comment by vinferrer** [comment]
okay I think this will do then

**Comment by eurunuela** [comment]
> okay I think this will do then

Then make it ready for review 😉

File: phys2bids/physio_obj.py [hunk 3]
```diff
@@ -7,6 +7,7 @@
 from itertools import groupby
 
 import numpy as np
+from copy import deepcopy
```
Comment (smoia) [inline phys2bids/physio_obj.py:10]: Could you move this import between logging and itertools please? It's a core module

**Review by smoia — CHANGES_REQUESTED** [review]
Execution looks good to me!

However, there's style-related point to address: can you fix the library to pass style checks and adjust the order of the imports in `physio_obj`?

I'm going to prepare all the rest to be able to merge this in.

**Review by eurunuela — APPROVED** [review]

**Comment by smoia** [comment]
Now that I think about it: @vinferrer could you make the `BlueprintOutput` initialisation similarly creating deepcopies of the inputs please?

**Comment by eurunuela** [comment]
Could you please fix the style? We already have a non-passing badge on the README, we should avoid a second one 😅

**Review by smoia — CHANGES_REQUESTED** [review]
Thank you @vinferrer !

My approval is dependent on passing all tests. Please fix the style issues.
(and possibly next time don't introduce them :wink:)

**Comment by vinferrer** [comment]
> Thank you @vinferrer !
> 
> My approval is dependent on passing all tests. Please fix the style issues.
> (and possibly next time don't introduce them 😉)

The changes required for the test to pass are in the PR #349, which is waiting for your review @smoia

**Comment by smoia** [comment]
> The changes required for the test to pass are in the PR #349, which is waiting for your review @smoia wink

Well, I don't know when we'll merge #349. So please fix them here so we don't propagate them any further.

**Review by smoia — APPROVED** [review]
LGTM!

**Comment by smoia** [comment]
(@tsalo if you're happy and you know it, merge this in! :clap: :clap: )

**Comment by smoia** [comment]
:rocket: PR was released in [`2.3.1`](https://github.com/physiopy/phys2bids/releases/tag/2.3.1) :rocket:

**Pull request merged**